eid_iomem_pci.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * Eidetic IOMEM Device Driver
+ * Copyright (c) 2022, Eideticom
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * Copyright (C) 2022 Eideitcom
+ */
+
+#include <linux/module.h>
+#include <linux/pci.h>
+#include <linux/pci-p2pdma.h>
+#include <linux/fs.h>
+#include <linux/cdev.h>
+#include <linux/pfn_t.h>
+
+#define PCI_VENDOR_EIDETICOM 0x1de5
+#define PCI_DEVICE_IOMEM 0x1000
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Andrew Maier <andrew.maier@eideticom.com>");
+MODULE_DESCRIPTION("Driver for the Eideticom IOMEM PCIe device");
+
+static int max_devices = 16;
+module_param(max_devices, int, 0444);
+MODULE_PARM_DESC(max_devices, "Maximum number of eid-iomem devices");
+
+static struct class *eid_iomem_class;
+static DEFINE_IDA(eid_iomem_ida);
+static dev_t eid_iomem_devt;
+
+static struct pci_device_id eid_iomem_pci_id_table[] = {
+	{ PCI_DEVICE(PCI_VENDOR_EIDETICOM, PCI_DEVICE_IOMEM), },
+	{ 0, }
+};
+MODULE_DEVICE_TABLE(pci, eid_iomem_pci_id_table);
+
+struct eid_iomem_dev {
+	struct device dev;
+	struct pci_dev *pdev;
+	int id;
+	struct cdev cdev;
+};
+
+static struct eid_iomem_dev *to_eid_iomem(struct device *dev)
+{
+	return container_of(dev, struct eid_iomem_dev, dev);
+}
+
+static int eid_iomem_open(struct inode *inode, struct file *filp)
+{
+	struct eid_iomem_dev *p;
+
+	p = container_of(inode->i_cdev, struct eid_iomem_dev, cdev);
+	filp->private_data = p;
+	//pci_p2pdma_file_open(p->pdev, filp);
+
+	return 0;
+}
+
+static int eid_iomem_mmap(struct file *filp, struct vm_area_struct *vma)
+{
+	struct eid_iomem_dev *p = filp->private_data;
+
+	return pci_mmap_p2pmem(p->pdev, vma);
+}
+
+static const struct file_operations eid_iomem_fops = {
+	.owner = THIS_MODULE,
+	.open = eid_iomem_open,
+	.mmap = eid_iomem_mmap,
+};
+
+static void eid_iomem_release(struct device *dev)
+{
+	struct eid_iomem_dev *p = to_eid_iomem(dev);
+
+	kfree(p);
+}
+
+static struct eid_iomem_dev *eid_iomem_create(struct pci_dev *pdev)
+{
+	struct eid_iomem_dev *p;
+	int err;
+
+	p = kzalloc(sizeof(*p), GFP_KERNEL);
+	if (!p)
+		return ERR_PTR(-ENOMEM);
+
+	p->pdev = pdev;
```

Comment:
Need to get and put a reference to the pdev....

The reference should be released in the eid_iomem_release so that it can still be used by the file operations.